src/fu-offline.c

```diff
@@ -16,51 +16,102 @@
 #include "fu-plugin-private.h"
 #include "fu-util-common.h"
 
+typedef enum {
+	FU_OFFLINE_FLAG_NONE		= 0,
+	FU_OFFLINE_FLAG_ENABLE		= 1 << 0,
+	FU_OFFLINE_FLAG_USE_PROGRESS	= 1 << 1,
+} FuOfflineFlag;
+
 struct FuUtilPrivate {
 	gchar		*splash_cmd;
 	GTimer		*splash_timer;
+	FuOfflineFlag	 splash_flags;
 };
 
 static gboolean
 fu_offline_set_splash_progress (FuUtilPrivate *priv, guint percentage, GError **error)
 {
-	g_autofree gchar *msg = NULL;
-	g_autofree gchar *str = NULL;
-	g_autoptr(GSubprocess) subprocess = NULL;
+	g_autofree gchar *str = g_strdup_printf ("%u", percentage);
+	const gchar *argv[] = { priv->splash_cmd, "system-update", "--progress", str, NULL };
+
+	/* call into plymouth if installed */
+	if (priv->splash_flags == FU_OFFLINE_FLAG_NONE) {
+		/* TRANSLATORS: console message when not using plymouth */
+		g_printerr ("%s: %u%%\n", _("Percentage complete"), percentage);
+		return TRUE;
+	}
+
+	/* fall back to really old mode that should be supported by anything */
+	if ((priv->splash_flags & FU_OFFLINE_FLAG_USE_PROGRESS) == 0) {
+		argv[1] = "display-message";
+		argv[2] = "--text";
+	}
+	return fu_common_spawn_sync (argv, NULL, NULL, 200, NULL, error);
+}
+
+static gboolean
+fu_offline_set_splash_mode (FuUtilPrivate *priv, GError **error)
+{
+	gboolean ret;
+	g_autoptr(GError) error_local = NULL;
+	const gchar *argv[] = { priv->splash_cmd, "change-mode", "--system-upgrade", NULL };
 
 	/* call into plymouth if installed */
-	if (priv->splash_cmd == NULL)
+	if (priv->splash_cmd == NULL) {
+		/* TRANSLATORS: console message when no Plymouth is installed */
+		g_printerr ("%s\n", _("Installing Firmware…"));
 		return TRUE;
+	}
+
+	/* try the new fancy mode */
+	ret = fu_common_spawn_sync (argv, NULL, NULL, 1500, NULL, &error_local);
+	if (ret) {
+		priv->splash_flags = FU_OFFLINE_FLAG_ENABLE;
+		priv->splash_flags = FU_OFFLINE_FLAG_USE_PROGRESS;
+		return TRUE;
+	}
+	g_debug ("failed to set system-upgrade mode, trying something else: %s",
```

Comment:
Rather than erroring, can you just check particular plymouth version this feature landed and set behavior based on that?

Reply:
Unfortunately not, Plymouth hasn't had a version bump in a long time. It's also dependant on the _runtime_ version rather than the _build-time_ version, so the two could easily get out of sync.

Reply:
OK. thanks for explaining.